fn.go

```diff
@@ -20,34 +34,200 @@ type Function struct {
-	// your Function. Input is arbitrary, except that it must be a KRM-like
-	// object. Supporting input is also optional - if you don't need to you can
-	// delete this, and delete the input directory.
 	in := &v1beta1.Input{}
 	if err := request.GetInput(req, in); err != nil {
 		response.Fatal(rsp, errors.Wrapf(err, "cannot get Function input from %T", req))
 		return rsp, nil
 	}
 
-	// TODO: Add your Function logic here!
-	//
-	// Take a look at function-sdk-go for some utilities for working with req
-	// and rsp - https://pkg.go.dev/github.com/crossplane/function-sdk-go
-	//
-	// Also, be sure to look at the tips in README.md
-	response.Normalf(rsp, "I was run with input %q", in.Example)
+	tg, err := NewTemplateSourceGetter(in)
+	if err != nil {
+		response.Fatal(rsp, errors.Wrap(err, fmt.Sprintf(errFmtInvalidFunction, errCannotGet)))
+		return rsp, nil
+	}
+
+	tmpl, err := GetNewTemplateWithFunctionMaps().Parse(tg.GetTemplates())
+	if err != nil {
+		response.Fatal(rsp, errors.Wrap(err, fmt.Sprintf(errFmtInvalidFunction, errCannotParse)))
+		return rsp, nil
+	}
+
+	reqMap, err := convertToMap(req)
+	if err != nil {
+		response.Fatal(rsp, errors.Wrap(err, "cannot convert request to map"))
+		return rsp, nil
+	}
+
+	f.log.Debug("constructed request map", "request", reqMap)
+
+	buf := &bytes.Buffer{}
+
+	if err := tmpl.Execute(buf, reqMap); err != nil {
+		response.Fatal(rsp, errors.Wrap(err, "cannot execute template"))
+		return rsp, nil
+	}
+
+	f.log.Debug("rendered manifests", "manifests", buf.String())
+
+	// Parse the rendered manifests.
+	var objs []*unstructured.Unstructured
+	decoder := yaml.NewYAMLOrJSONDecoder(bytes.NewBufferString(buf.String()), 1024)
+	for {
+		u := &unstructured.Unstructured{}
+		if err := decoder.Decode(&u); err != nil {
+			if err == io.EOF {
+				break
+			}
+			response.Fatal(rsp, errors.Wrap(err, "cannot decode manifest"))
+			return rsp, nil
+		}
+		if u != nil {
+			objs = append(objs, u)
+		}
+	}
+
+	// Get the desired composite resource from the request.
+	desiredComposite, err := request.GetDesiredCompositeResource(req)
+	if err != nil {
+		response.Fatal(rsp, errors.Wrap(err, "cannot get desired composite resource"))
+		return rsp, nil
+	}
+
+	//  Get the desired composed resources from the request.
+	desiredComposed, err := request.GetDesiredComposedResources(req)
+	if err != nil {
+		response.Fatal(rsp, errors.Wrap(err, "cannot get desired composed resources"))
+		return rsp, nil
+	}
+
+	// Convert the rendered manifests to a list of desired composed resources.
+	for _, obj := range objs {
+		cd := resource.NewDesiredComposed()
+		cd.Resource.Unstructured = *obj.DeepCopy()
+
+		// TODO(ezgidemirel): Refactor to reduce cyclomatic complexity.
+		// Update only the status of the desired composite resource.
+		if cd.Resource.GetAPIVersion() == desiredComposite.Resource.GetAPIVersion() && cd.Resource.GetKind() == desiredComposite.Resource.GetKind() {
```

Comment:
In what situation would a desired composed resource have the same GVK as the desired composite resource (XR)?

Reply:
We may add the composite resource in our templates to set a status field. I extended examples/composition.yaml file with an example.

Reply:
Ah, got it. Makes sense!